common/src/main/java/org/apache/celeborn/common/network/client/TransportClientFactory.java

```diff
@@ -250,10 +258,31 @@ public void initChannel(SocketChannel ch) {
     }
 
     TransportClient client = clientRef.get();
+    Channel channel = channelRef.get();
     assert client != null : "Channel future completed successfully with null client";
 
+    // Execute any client bootstraps synchronously before marking the Client as successful.
+    long preBootstrap = System.nanoTime();
+    logger.debug("Running bootstraps for {} ...", address);
+    try {
+      for (TransportClientBootstrap clientBootstrap : clientBootstraps) {
+        clientBootstrap.doBootstrap(client, channel);
+      }
+    } catch (Exception e) { // catch non-RuntimeExceptions too as bootstrap may be written in Scala
+      long bootstrapTime = System.nanoTime() - preBootstrap;
+      logger.error(
+          "Exception while bootstrapping client after {}",
+          Utils.nanoDurationToString(bootstrapTime),
+          e);
+      client.close();
```

Comment:
Use try-with-resources to avoid client not being closed.

Reply:
It is not managing an `AutoCloseable` resource here. We are iterating over clientbootstraps and if there is any exception during bootstrap, we close the client.

Reply:
@otterc, if there is no exception, does the client need to close?

Reply:
No, that means the connection was successfully bootstrapped and it is returned to the caller.